### PR TITLE
Pwcnet dev

### PR DESCRIPTION
**Pull request by NeedsMorePie**
Optimized input pipeline IO by reducing file sizes (bytes instead of floats, added compression).
Optimized input pipeline shuffling by shuffling just the files.
Working flyingthings dataset.
Mechanism to auto-run the pwcnet training pipeline with all the config and dataset switches.

### CODE REVIEW

File: data/flow/flowdata.py [hunk 1]
```diff
@@ -281,36 +310,47 @@ def _parse_function(example_proto):
...
+    :param max_flow: Float. Maximum flow magnitude of the flow image. Any examples with flow magnitude greater than this
+        will be ignored.
     :return: Nothing.
     """
     if verbose and len(shard_range) > 0:
         print('Writing to shard', shard_id, 'data points', shard_range[0], 'to', shard_range[-1])
 
-    writer = tf.python_io.TFRecordWriter(os.path.join(directory, str(shard_id) + '_' + filename))
+    record_name = os.path.join(directory, str(shard_id) + '_' + filename)
+    options = tf.python_io.TFRecordOptions(tf.python_io.TFRecordCompressionType.GZIP)
```
Comment (boxofpasta) [inline data/flow/flowdata.py]: How are the savings on disk with this compression, roughly speaking?

File: data/flow/flowdata.py [hunk 2]
```diff
@@ -33,6 +34,8 @@ def __init__(self, directory, batch_size=1, validation_size=1, crop_size=None, t
         :param training_augmentations: Whether to do live augmentations while training.
         :param data_source: Source of the data.
         :param augmentation_config: Configurations for data augmentation. If None, the default will be used.
+        :param max_flow: Float. Maximum flow magnitude of the flow image. Any examples with flow magnitude greater than
```
Comment (boxofpasta) [inline data/flow/flowdata.py]: What is this needed for?

File: utils/misc.py [hunk 3]
```diff
@@ -43,3 +44,65 @@ def print_progress_bar(iteration, total, prefix= '', suffix='', decimals=1, leng
     # Print New Line on Complete
     if iteration == total:
         print()
+
+
+def preprocess_var_refs(root, base_vars=None, remove_vars_dict=True):
+    """
+    Processes the dict by replacing var refs with variables declared in the 'vars' dict.
```
Comment (boxofpasta) [inline utils/misc.py]: These comments are not descriptive enough, maybe add a simple example (one of the unit tests will do).

File: utils/misc.py [hunk 4]
```diff
@@ -43,3 +44,65 @@ def print_progress_bar(iteration, total, prefix= '', suffix='', decimals=1, leng
...
+    :param root: Dict.
+    :param vars: Dict. Vars to add to the dict.
```
Comment (boxofpasta) [inline utils/misc.py]: Should this be base_vars?

File: utils/misc_test.py [hunk 5]
```diff
@@ -0,0 +1,166 @@
...
+            },
+            "boop": [
+                "bar",
+                2
+            ]
+        }
+        """
+        self.json_equals(json_str, expected_json_str)
+
+    def json_equals(self, json_str, expected_json_str):
```
Comment (boxofpasta) [inline utils/misc_test.py]: This function name is misleading.

File: utils/tf_test.py [hunk 6]
```diff
@@ -54,3 +52,7 @@ def test_sliding_window_slice_small(self):
         sliced = sliding_window_slice(x, slice_locations)
         sliced_list = self.sess.run(sliced).tolist()
         self.assertListEqual(sliced_list, expected)
+
+
+if __name__ == '__main__':
```
Comment (boxofpasta) [inline utils/tf_test.py]: Woops, thanks.

File: data/flow/flowdata.py (hunk 1 above)
Comment (NeedsMorePie) [reply data/flow/flowdata.py]: Good savings. 10-20% disk space saved. Given that we are working with shared network drives in the cluster, I believe that this helps.

File: data/flow/flowdata.py (hunk 2 above)
Comment (NeedsMorePie) [reply data/flow/flowdata.py]: It said in the paper to discard examples with flow greater than 1000. In my debugging print statements, there are actually flyingthings examples with flows larger than this.

File: utils/misc.py (hunk 4 above)
Comment (NeedsMorePie) [reply utils/misc.py]: Yup.

File: utils/misc_test.py (hunk 5 above)
Comment (NeedsMorePie) [reply utils/misc_test.py]: Yup.

**Pull request merged**